### PR TITLE
[DTensor] fix scaled_mm sharding strategy

### PR DESCRIPTION
**Pull request by pianpwk**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #177235
* __->__ #177234

For https://github.com/pytorch/ao/pull/4010, the existing strategy incorrect copies scaled_mm input strategies (2d) onto the scale tensor (1d)

### CODE REVIEW

File: torch/distributed/tensor/_ops/_matrix_ops.py [hunk 1]
```diff
@@ -202,6 +249,19 @@ def _scaled_mm_like_strategy(
         raise AssertionError("_scaled_mm on DTensors doesn't support bias")
     if scale_result_strategy is not None:
         raise AssertionError("_scaled_mm on DTensors doesn't support scale_result")
+
+    # Derive contracting dimensions from the einsum equation.
+    # For "mk,kn->mn": contracting char is 'k', self_dim=1, mat2_dim=0.
+    inputs, output = mm_equation.split("->")
+    input_terms = inputs.split(",")
```
Comment (Skylion007) [inline torch/distributed/tensor/_ops/_matrix_ops.py]: May want to add an insertion on the number of input terms?

File: test/distributed/tensor/test_matrix_ops.py [hunk 2]
```diff
@@ -527,6 +528,67 @@ def test_scaled_mm(self):
...
+        result = _scaled_mm_scale_spec(spec_shard0, torch.Size([64]), contracting_dim=0)
+        self.assertIsNone(result)
+
+        # --- 1D blockwise + Replicate -> Replicate ---
+        spec_repl = DTensorSpec(device_mesh, (Replicate(),))
+        result = _scaled_mm_scale_spec(spec_repl, torch.Size([64]), contracting_dim=1)
+        self.assertIsNotNone(result)
+        self.assertEqual(result.placements, (Replicate(),))
+
+        # --- 1D blockwise + Partial -> Replicate ---
```
Comment (danielvegamyhre) [inline test/distributed/tensor/test_matrix_ops.py:586]: can you explain this partial -> replicate rule for blockwise scales? wouldn't this indicate an all-reduce?

File: torch/distributed/tensor/_ops/_matrix_ops.py [hunk 3]
```diff
@@ -177,6 +177,53 @@ def _addmm_like_strategy(
...
+
+    if len(scale_shape) != 1:
+        # 2D (or higher) scale: copy data spec directly (e.g. row-wise)
+        return data_spec
+
+    # 1D blockwise scale: Shard(>=1) is invalid on a 1D tensor, so we need
+    # to map the data operand's placement to a valid 1D placement.
+    new_placements: list[Placement] = []
+    for p in data_spec.placements:
+        if isinstance(p, Replicate | Partial):
```
Comment (wconstab) [inline torch/distributed/tensor/_ops/_matrix_ops.py]: hmm, seems like at least this helper should be written in a single-mesh dim way, since ultimately we want to replace the strategy_fn too

File: torch/distributed/tensor/_ops/_matrix_ops.py [hunk 4]
```diff
@@ -202,6 +249,19 @@ def _scaled_mm_like_strategy(
         raise AssertionError("_scaled_mm on DTensors doesn't support bias")
     if scale_result_strategy is not None:
         raise AssertionError("_scaled_mm on DTensors doesn't support scale_result")
+
+    # Derive contracting dimensions from the einsum equation.
+    # For "mk,kn->mn": contracting char is 'k', self_dim=1, mat2_dim=0.
+    inputs, output = mm_equation.split("->")
```
Comment (wconstab) [inline torch/distributed/tensor/_ops/_matrix_ops.py]: we have a helper in einsum_strategy.py that does some of this, can it be reused?

File: torch/distributed/tensor/_ops/_matrix_ops.py [hunk 5]
```diff
@@ -214,19 +274,18 @@ def _scaled_mm_like_strategy(
...
         mat2_spec = strtg.input_specs[1]
-        # propagate the operands' specs to their scales, except for tensor-wise
-        # scaling which can have any numbers of dims (legacy...), hence sharding
-        # dims won't map. for tensor-wise, anyways, we can only do replication.
-        scale_self_spec = (
-            DTensorSpec(self_spec.mesh, (Replicate(),))
-            if prod(scale_self_strategy.shape) == 1
-            else self_spec
+        # Propagate the operands' specs to their scales. For tensor-wise
+        # scales we always replicate. For 2D scales (e.g. row-wise) we copy
```
Comment (wconstab) [inline torch/distributed/tensor/_ops/_matrix_ops.py]: could you add to the PR desc a bit more detail about the different scale formats we're supporting and how we shard them?

i'm confused about where a 1D scale comes from. if it's blockwise according to MX format, i would expect that the scale tensor is the same shape as the data tensor, except divided by 32 in the dimension we're scaling. this would still yield a 2D scale tensor in this case wouldn't it?

Comment (danielvegamyhre) [reply torch/distributed/tensor/_ops/_matrix_ops.py]: the mxfp8 scale factors are weird. for context, how we get 1d scales from a 2d tensor of shape (M,K):
- initial scale shape: (M, K//32) in row major layout
- tcgen05.mma PTX instruction requires scale factors be in a special hierarchical ((32,4),4) layout for consumption by tensor cores. therefore, we transform each 128x4 scale factor tile from row major layout, into a contiguous 512 byte chunk in ((32,4),4) layout. 
- Shape is then flattened to a 1d tensor that just acts as a buffer containing contiguous 512 byte scale factors, in "row of tiles" major order. 

Our recent blog has some diagrams of these layouts for grouped GEMMs if you're interested: https://pytorch.org/blog/mxfp8-training-for-moes-1-3x-training-speedup-vs-bf16-for-llama4-scout-on-gb200-cluster-using-torchao-and-torchtitan/

File: test/distributed/tensor/test_matrix_ops.py (hunk 2 above)
Comment (pianpwk) [reply test/distributed/tensor/test_matrix_ops.py:586]: I think this means the matmul is PxR->P or RxP->P, with replicated scale tensors. You can delay the allreduce until after, though this rule might never be used?

File: test/distributed/tensor/test_matrix_ops.py [hunk 6]
```diff
@@ -527,6 +527,69 @@ def test_scaled_mm(self):
...
+        )
+        self.assertIsNotNone(result)
+        self.assertEqual(result, Shard(0))
+
+        # B_t (kn): dim 1 = n (non-contracting), dim 0 = k (contracting)
+        result = _scaled_mm_scale_placement(
+            Shard(1), torch.Size([64]), contracting_dim=0
+        )
+        self.assertIsNotNone(result)
+        self.assertEqual(result, Shard(0))
```
Comment (danielvegamyhre) [inline test/distributed/tensor/test_matrix_ops.py:567]: for B_t sharded along dim1, how does the output of A@B_t become sharded along dim0? i am picturing something like this:

Image

File: torch/distributed/tensor/_ops/_matrix_ops.py [hunk 7]
```diff
@@ -177,83 +177,41 @@ def _addmm_like_strategy(
...
+
+    # 1D blockwise scale: Shard(>=1) is invalid on a 1D tensor, so we need
+    # to map the data operand's placement to a valid 1D placement.
+    if isinstance(data_placement, _ShardingPlaceholder):
+        if data_placement.dim == contracting_dim:
+            return None
+        return _ShardingPlaceholder(0)
+    elif isinstance(data_placement, Shard):
+        if data_placement.dim == contracting_dim:
+            return None
```
Comment (danielvegamyhre) [inline torch/distributed/tensor/_ops/_matrix_ops.py:210]: does returning None in these rules basically indicate the placement is invalid? will it crash / return a user facing error message?

File: test/distributed/tensor/test_matrix_ops.py (hunk 6 above)
Comment (pianpwk) [reply test/distributed/tensor/test_matrix_ops.py:567]: hmm I think S(0) is the corresponding scale tensor's placement - this is saying for an input (2d matrix, 1d scale tensor), if the matrix is Shard(1), the scale tensor will be sharded (but Shard(0) because it's 1d)

File: torch/distributed/tensor/_ops/_matrix_ops.py (hunk 7 above)
Comment (pianpwk) [reply torch/distributed/tensor/_ops/_matrix_ops.py:210]: nah it just doesn't enumerate the placement as a valid option

File: torch/distributed/tensor/_ops/_matrix_ops.py [hunk 8]
```diff
@@ -489,10 +447,36 @@ def baddbmm_single_dim_strategy(
...
+        # strat is [output, self, mat2]; derive scale placements
+        scale_self_p = _scaled_mm_scale_placement(
+            strat[1], scale_self_meta.shape, contracting_dim=1
+        )
+        scale_mat2_p = _scaled_mm_scale_placement(
+            strat[2], scale_mat2_meta.shape, contracting_dim=0
+        )
+        if scale_self_p is None or scale_mat2_p is None:
+            continue
+        result.append(strat + [scale_self_p, scale_mat2_p])
```
Comment (danielvegamyhre) [inline torch/distributed/tensor/_ops/_matrix_ops.py:478]: for my own understanding: 
- `scaled_mm_single_dim_strategy` is generating custom sharding rules for when input(s) are sharded over a single dim of a mesh (what will happen if they are sharded over 2 dims, like FSDP + TP?)
- `gen_single_dim_einsum_strategies` just enumerates all the placement possibilities, i.e. `Shard(0) @ Replicate -> Shard(0)`, `Shard(1) @ Replicate -> Shard(1)` etc for the operation (matmul in this case)
- Then you are adding the scale_a/scale_b placements for each rule, based on custom logic for the 1d scales.

Is that right?

**Review by danielvegamyhre — APPROVED** [review]

**Comment by danielvegamyhre** [comment]
LGTM, left a couple questions for my own understanding. You may want final review from someone with more dtensor expertise, up to you. thanks again for working on this!

File: torch/distributed/tensor/_ops/_matrix_ops.py (hunk 5 above)
Comment (wconstab) [reply torch/distributed/tensor/_ops/_matrix_ops.py]: so i think i understand how we can shard the flattened 1d tensor S(0) and have that work with a data-tensor that is also S(0) sharded.  I am not sure any other sharding combination should be possible though.  

And for S0 with S0, I think we also have to ensure that the data tensor local shard dim size is a multiple of 128, don't we?  Because if for example we only have one row of the data tensor per rank, then there is not a valid way to shard the scale tensor accordingly right?

Comment (danielvegamyhre) [reply torch/distributed/tensor/_ops/_matrix_ops.py]: >And for S0 with S0, I think we also have to ensure that the data tensor local shard dim size is a multiple of 128, don't we? Because if for example we only have one row of the data tensor per rank, then there is not a valid way to shard the scale tensor accordingly right?

When we write scales to blocked layout, the rows/cols are [padded](https://github.com/pytorch/ao/blob/eb64bfbd2599350e3f6ce5640d4b3a748feee997/torchao/prototype/mx_formats/utils.py#L51-L57) such that they are evenly divisible into 128x4 chunks. (this is the pytorch native impl, but our custom kernels do this as well). So I don't think we have to enforce shard size constraints.

File: torch/distributed/tensor/_ops/_matrix_ops.py (hunk 8 above)
Comment (pianpwk) [reply torch/distributed/tensor/_ops/_matrix_ops.py:478]: > scaled_mm_single_dim_strategy is generating custom sharding rules for when input(s) are sharded over a single dim of a mesh (what will happen if they are sharded over 2 dims, like FSDP + TP?)

> Then you are adding the scale_a/scale_b placements for each rule, based on custom logic for the 1d scales.

yep!

> gen_single_dim_einsum_strategies just enumerates all the placement possibilities, i.e. Shard(0) @ Replicate -> Shard(0), Shard(1) @ Replicate -> Shard(1) etc for the operation (matmul in this case)

we combinatorially enumerate all possible strategies over each mesh dim (they can be independent), then the redistribution cost b/w real input placements <-> each of the 2d mesh placement on (FSDP, TP) is calculated, and the lowest cost one chosen
